### PR TITLE
20-Squad Kai Educator Tools Output History

### PR DESCRIPTION
**Pull request by Cosmodocus**
# Description
This is Squad-20's Kai Educator History Tools Output feature created and implemented onto the already existing Kai Project Application. The features involve the user utilizing the Kai Tools(Multiple Choice or Flash Cards), having the data saved and displayed in the History page, seeing the saved data in more detail displayed on a side drawer after clicking preview, and being able to copy the content as well as export the content as a CSV file. The data stored is also filtered out by date in a similar fashion to ChatGPT's sidebar. 

Demo Videos:
a) Detailed Explanation of the Backend by Yichen Wang:
https://www.loom.com/share/6569bdf12710454387ab1786a85411a5?sid=572afd77-e0d0-4598-ab74-6324f63375b7

b) Full Feature Demo of Kai Educational Tools History Output by Reza Homayuni(recommended to watch in 1.5x speed):
https://www.loom.com/share/382649f2b8494b7a9d326e11a4ed0de4?sid=13493708-590a-465b-a364-74e879fbdc37

Database Schema:
- Multiple Choice Assessment Database Schema: https://app.eraser.io/workspace/ucwIyZfNQAPhvvIZJ8u6
- Youtube Flash Cards Database Schema: https://app.eraser.io/workspace/1lmWKcg1l45TMZtR0yBh

Test Link:
https://kai-project-history.web.app/

Ticket:

## Type of change
Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration. Include screenshot/video if possible.
- [x]  Test A - Verified tool functionality using YouTube video links to generate quizzes, ensuring correct display of tool data including date, title, and information.
- [x]  Test B - Confirmed proper functionality and sorting by dates during tool loading.
- [x] Test C - Checked the created card using the preview button to ensure  button functionality and correct display across different resolutions, and proper display of the slide panel with available tools.
- [x] Test D - Tested copy button functionality by pasting data into NotePad or Microsoft Word to verify correct display of multiple choice quiz and flash cards content.
- [x]  Test E - Verified export button functionality, ensuring data exported correctly to expected formats (MCQ or flash card) with a table containing tool data in the exported file.
- [x]  Test F - We minimized delays between tools and actions and ensured consistent chat history output across various resolutions and browsers (Safari, Edge, Chrome, Firefox).

### CODE REVIEW

File: frontend/services/tools/submitPrompt.js [hunk 1]
```diff
@@ -1,29 +1,59 @@
...
       },
     });
 
+    const toolSessionPayload = {
+      user: payload.user,
+      tool_data: { ...payload.tool_data },
+      type: payload.type,
+      messages: response.data,
+      sessionId: payload.sessionId,
+    };
```
Comment (bkb-Git) [inline frontend/services/tools/submitPrompt.js:33]: Why are we doing this here ?

File: frontend/utils/FirebaseUtils.jsx [hunk 2]
```diff
@@ -1,30 +0,0 @@
...
-    unixTimestamp = moment(date.toDate()).valueOf();
-  } else {
-    // Invalid or unsupported date value
-    unixTimestamp = null;
-  }
-
-  return unixTimestamp;
-};
-
-export { parseFirebaseText, convertToUnixTimestamp };
```
Comment (bkb-Git) [inline frontend/utils/FirebaseUtils.jsx:30]: What's going on here ?

File: frontend/utils/MiscellaneousUtils.js [hunk 3]
```diff
@@ -198,6 +199,148 @@ const getRandomBackgroundColor = () => {
...
+ * @param {Object} item - The item containing tool data.
+ * @returns {Object} Processed tool card data.
+ */
+const getToolCardData = (item) => {
+  const toolId = item.tool_data?.[0]?.tool_id;
+  const handler = ToolCardHandlers[toolId] || ToolCardHandlers.default;
+
+  return handler(item);
+};
+
```
Comment (bkb-Git) [inline frontend/utils/MiscellaneousUtils.js:343]: Do we need this logic defined here, it seems pretty specific ?

File: frontend/utils/TimeUtils.jsx [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+  return (
+    <HistoryListingContainer
+      key={category}
+      data={timeData}
+      loading={loading}
+      category={category}
+    />
+  );
+};
```
Comment (bkb-Git) [inline frontend/utils/TimeUtils.jsx:51]: Don't think this should be here

File: frontend/utils/TimeUtils.jsx [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+      {renderHistoryToolContainer('Today', today, loading)}
+      {renderHistoryToolContainer('Yesterday', yesterday, loading)}
+      {renderHistoryToolContainer('Previous 7 days', previous7Days, loading)}
+      {renderHistoryToolContainer('Previous 30 days', previous30Days, loading)}
+      {Object.entries(monthsBefore).map(([month, timeData]) =>
+        renderHistoryToolContainer(month, timeData, loading)
+      )}
+    </>
+  );
+};
```
Comment (bkb-Git) [inline frontend/utils/TimeUtils.jsx:75]: This too, this shouldn't be here

**Review by bkb-Git — CHANGES_REQUESTED** [review]
Left a couple of comments that need to be resolved. Let me know if you have any questions.

File: frontend/layouts/MainAppLayout/NavMenu/NavMenu.jsx [hunk 6]
```diff
@@ -35,25 +42,30 @@ const NavMenu = () => {
...
     router.push(link);
-    setActive(id);
   };
 
   return (
     <Grid {...styles.mainGridProps}>
       {PAGES.map((page) => (
         <MenuItem
           key={page.id}
-          onClick={() => handleRoute(page.link, page.id)}
```
[the remark this answers is not in the document]
Comment (Cosmodocus) [reply frontend/layouts/MainAppLayout/NavMenu/NavMenu.jsx:56]: Yes it still does. I've tested it with random url routes, and they all redirect to the home page if it doesn't exist.

File: frontend/hooks/useFilterByTime.jsx
[the remark this answers is not in the document]
Comment (Cosmodocus) [reply frontend/hooks/useFilterByTime.jsx:1]: Thank you

File: frontend/services/tools/submitPrompt.js (hunk 1 above)
Comment (wyichen99) [reply frontend/services/tools/submitPrompt.js:33]: We got response data (output data) in line 21 stored in response. And I stored the output as messages(I will change it to outputs), and send to firestore using createtoolsession function.

**Review by bkb-Git — APPROVED** [review]
LGTM!

**Pull request merged**